1-Draft/RFCNNNN-#Requires-Additions.md

```diff
@@ -0,0 +1,155 @@
+  in force.
+* Using `#requires` in the interactive console will cause
+  a parse-time error. This could be a **minor breaking
+  change**, since currently PowerShell throws a [pipeline
+  creation error](https://github.com/PowerShell/PowerShell/issues/3803).
+* Add support for the following new parameters (each
+  independently up for discussion):
+  * `-OS {Windows | Linux | MacOS}`, where an
+    operating system (or possibly combination of them) can
+    be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/3751).
+  * `-Assembly <Assembly-name>`, where a .NET assembly can
+    be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/5022).
+  * `-MaxVersion <V>[.<v>]`, where a maximum PowerShell
+     version can be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/2846).
+
+## Motivation
+
+> As a PowerShell user, I can be sure that all the
+> `#requires` statements in a script come before any
+> PowerShell code, so that it's clear they always
+> execute first, and so they can all be found easily.
+
+> As a PowerShell user, I get feedback that `#requires`
+> statements cannot be used in the interactive console,
+> so that it's clear that they have no effect on an
+> interactive session.
+
+> As a PowerShell user, I can specify that my script
+> `#requires` being run on a specific operating system,
+> so that I can effectively, efficiently and declaratively
+> guarantee that it is used only on systems I designed it
+> for.
+
+> As a PowerShell user, I can specify that my script
+> `#requires` a given .NET Assembly to run in an
+> efficient and declarative way, so that I don't have
+> to do complex runtime logic to determine my script
+> cannot run.
+
+> As a PowerShell user, I can specify that my script
+> `#requires` to be run in a version of PowerShell
+> lower than a given version, so that I can declaratively
+> prevent it from being run in an environment where
+> changes to PowerShell would cause unintended behavior.
+
+## Specification
+
+1. `#requires` statements must appear in scripts
+   above all executable PowerShell. Any `#requires`
+   statement placed after any PowerShell code causes
+   an unrecoverable parse-time error. This new restriction
+   should not affect the usage of other comment-embedded
+   directives or pragmas, such as `#sig`, linter directives or
+   inline editor configurations. Specifically, the new `#requires`
+   placement restriction must not interfere with Unix-style
+   hashbangs (e.g. `#!/usr/bin/pwsh`).
+2. Any use of `#requires` in an interactive session causes
+   a specific parse-time error to be thrown, informing the
+   user that `#requires` may not be used in the interactive
+   console.
+3. `#requires` can take an `-OS` parameter, with possible
+   arguments being `Windows`, `Linux` and `MacOS` (and
+   possibly some syntax for `or`-ing them). This check
```

Comment:
Probably `-OS Linux,macOS`

Reply:
Maybe use generic `-OS Unix` for al supported Unixes?
And should we consider OS version? `-OS Unix(Ubuntu 16.04)`

Reply:
Using `Unix` as generic term for `Linux` , `Unix`, and `macOS` is 1) not accurate and 2) really makes some people angry. 🙁 I would advise against anything Microsoft does making that mistake as it then makes MS look either ignorant or arrogant depending on what issue a person takes with that usage.

Reply:
@markekraus: I think @iSazonov is suggesting `Unix` as an _additional_ identifier, to serve as an _abstraction_ over macOS and Linux for scripts that run on _both_. Given that both platforms have a Unix _heritage_ and many commonalities (especially in contrast with Windows), it is a useful abstraction. It is the case variant `UNIX` that is trademarked; I think that the term `Unix` to loosely refer to all platforms with a Unix heritage is well-established - see https://en.wikipedia.org/wiki/Unix

Reply:
@mklement0 Yes, and that usage is contentious to the point of being inflammatory. That's is why I'm recommending against it. I understood @iSazonov intent. 

I'll also counter that it is not practical. My exposure to CoreFX is that macOS is it's own beast. That's not even considering the radically different ecosystem from Linux. The gap is wide enough that I see very few cases where Unix would be useful to mean `not-windows`.

With that and the inflammatory nature a Microsoft product of using `Unix` to lump them together further coupled with "there is nothing wrong with `-OS Linux,macOS`" I'd say we shouldn't use it.

Though another category might be helpful: `Windows`, `Linux`, `macOS`, `IoT`

Reply:
@markekraus: Whether the abstraction is _useful_ should be discussed separately from its _name_.

* To me, the abstraction is useful (speaking as a shell user on both platforms) - it all depends on the use case; perhaps others can weigh in.

* If the name `Unix` is really that contentious (which I'm personally not aware of), perhaps `UnixLike` will do.

Reply:
Current .Net Core team support goal is "we should run on all platforms". So maybe we should use more generic parameter `-OS` and sub-parameters for platform, version ...